File: src/specforge_api/routers/quality.py
"""FastAPI router for Quality Guardian (Module 7).

Endpoints
---------
GET    /api/v1/projects/{project_id}/quality/checklists
POST   /api/v1/projects/{project_id}/quality/checklists
GET    /api/v1/projects/{project_id}/quality/checklists/{checklist_id}
PUT    /api/v1/projects/{project_id}/quality/checklists/{checklist_id}/items/{item_id}
DELETE /api/v1/projects/{project_id}/quality/checklists/{checklist_id}
GET    /api/v1/projects/{project_id}/quality/test-suites
POST   /api/v1/projects/{project_id}/quality/test-suites
GET    /api/v1/projects/{project_id}/quality/test-suites/{suite_id}
GET    /api/v1/projects/{project_id}/quality/reports
POST   /api/v1/projects/{project_id}/quality/reports

NOTE: Static sub-paths (test-suites, reports) are registered BEFORE the
parameterised {checklist_id} / {suite_id} routes so FastAPI matches them
correctly.
"""

from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, status

from specify_cli.quality import (
    AnalysisFinding,
    AnalysisReport,
    Checklist,
    ChecklistItem,
    ChecklistItemStatus,
    SeverityLevel,
    TestCase,
    TestCaseType,
    TestSuite,
)
from specforge_api.config import Settings, settings as default_settings
from specforge_api.schemas.quality import (
    AnalysisFindingSchema,
    AnalysisReportCreateRequest,
    AnalysisReportListResponse,
    AnalysisReportSchema,
    ChecklistCreateRequest,
    ChecklistItemSchema,
    ChecklistItemUpdateRequest,
    ChecklistListResponse,
    ChecklistSchema,
    TestCaseSchema,
    TestSuiteCreateRequest,
    TestSuiteListResponse,
    TestSuiteSchema,
)

router = APIRouter()


def _get_settings() -> Settings:
    return default_settings


# ---------------------------------------------------------------------------
# Path helpers
# ---------------------------------------------------------------------------


def _checklists_dir(project_id: str, projects_root: Path) -> Path:
    return projects_root / project_id / "quality" / "checklists"


def _checklist_path(project_id: str, checklist_id: str, projects_root: Path) -> Path:
    return _checklists_dir(project_id, projects_root) / f"{checklist_id}.yaml"


def _suites_dir(project_id: str, projects_root: Path) -> Path:
    return projects_root / project_id / "quality" / "test-suites"


def _suite_path(project_id: str, suite_id: str, projects_root: Path) -> Path:
    return _suites_dir(project_id, projects_root) / f"{suite_id}.yaml"


def _reports_dir(project_id: str, projects_root: Path) -> Path:
    return projects_root / project_id / "quality" / "reports"


def _report_path(project_id: str, report_id: str, projects_root: Path) -> Path:
    return _reports_dir(project_id, projects_root) / f"{report_id}.yaml"


# ---------------------------------------------------------------------------
# Load-or-404 helpers
# ---------------------------------------------------------------------------


def _load_checklist_or_404(project_id: str, checklist_id: str, cfg: Settings) -> Checklist:
    path = _checklist_path(project_id, checklist_id, cfg.projects_root)
    try:
        return Checklist.load(path)
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Checklist '{checklist_id}' not found in project '{project_id}'",
        )


def _load_suite_or_404(project_id: str, suite_id: str, cfg: Settings) -> TestSuite:
    path = _suite_path(project_id, suite_id, cfg.projects_root)
    try:
        return TestSuite.load(path)
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"TestSuite '{suite_id}' not found in project '{project_id}'",
        )


def _load_report_or_404(project_id: str, report_id: str, cfg: Settings) -> AnalysisReport:
    path = _report_path(project_id, report_id, cfg.projects_root)
    try:
        return AnalysisReport.load(path)
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"AnalysisReport '{report_id}' not found in project '{project_id}'",
        )


# ---------------------------------------------------------------------------
# Conversion helpers: domain → schema
# ---------------------------------------------------------------------------


def _item_schema(item: ChecklistItem) -> ChecklistItemSchema:
    return ChecklistItemSchema(
        item_id=item.item_id,
        checklist_id=item.checklist_id,
        category=item.category,
        description=item.description,
        status=item.status.value,
        notes=item.notes,
    )


def _checklist_schema(checklist: Checklist) -> ChecklistSchema:
    return ChecklistSchema(
        checklist_id=checklist.checklist_id,
        project_id=checklist.project_id,
        name=checklist.name,
        spec_id=checklist.spec_id,
        created_at=checklist.created_at,
        items=[_item_schema(i) for i in checklist.items],
    )


def _case_schema(case: TestCase) -> TestCaseSchema:
    return TestCaseSchema(
        case_id=case.case_id,
        suite_id=case.suite_id,
        title=case.title,
        case_type=case.case_type.value,
        scenario=case.scenario,
        expected=case.expected,
        tags=case.tags,
    )


def _suite_schema(suite: TestSuite) -> TestSuiteSchema:
    return TestSuiteSchema(
        suite_id=suite.suite_id,
        project_id=suite.project_id,
        feature=suite.feature,
        spec_id=suite.spec_id,
        generated_at=suite.generated_at,
        cases=[_case_schema(c) for c in suite.cases],
    )


def _finding_schema(finding: AnalysisFinding) -> AnalysisFindingSchema:
    return AnalysisFindingSchema(
        finding_id=finding.finding_id,
        report_id=finding.report_id,
        severity=finding.severity.value,
        category=finding.category,
        description=finding.description,
        recommendation=finding.recommendation,
    )


def _report_schema(report: AnalysisReport) -> AnalysisReportSchema:
    return AnalysisReportSchema(
        report_id=report.report_id,
        project_id=report.project_id,
        spec_id=report.spec_id,
        plan_id=report.plan_id,
        analyzed_at=report.analyzed_at,
        findings=[_finding_schema(f) for f in report.findings],
    )


# ===========================================================================
# Checklist endpoints
# ===========================================================================

# ---------------------------------------------------------------------------
# List checklists  — static, must precede parameterised routes
# ---------------------------------------------------------------------------


@router.get(
    "/projects/{project_id}/quality/checklists",
    response_model=ChecklistListResponse,
    summary="List all quality checklists for a project",
)
def list_checklists(
    project_id: str,
    cfg: Settings = Depends(_get_settings),
) -> ChecklistListResponse:
    """Return all checklists for *project_id* (empty list if none exist yet)."""
    checklists_dir = _checklists_dir(project_id, cfg.projects_root)
    checklists: list[ChecklistSchema] = []
    if checklists_dir.exists():
        for yaml_file in sorted(checklists_dir.glob("*.yaml")):
            try:
                checklist = Checklist.load(yaml_file)
                checklists.append(_checklist_schema(checklist))
            except Exception:
                continue
    return ChecklistListResponse(project_id=project_id, checklists=checklists)


# ---------------------------------------------------------------------------
# Create checklist
# ---------------------------------------------------------------------------


@router.post(
    "/projects/{project_id}/quality/checklists",
    response_model=ChecklistSchema,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new quality checklist",
)
def create_checklist(
    project_id: str,
    body: ChecklistCreateRequest,
    cfg: Settings = Depends(_get_settings),
) -> ChecklistSchema:
    """Create and persist a new Checklist for *project_id*."""
    checklist = Checklist(
        project_id=project_id,
        name=body.name,
        created_at=body.created_at,
        spec_id=body.spec_id,
    )
    for item_req in body.items:
        item = ChecklistItem(
            checklist_id=checklist.checklist_id,
            category=item_req.category,
            description=item_req.description,
            notes=item_req.notes,
        )
        checklist.add_item(item)
    path = _checklist_path(project_id, checklist.checklist_id, cfg.projects_root)
    checklist.save(path)
    return _checklist_schema(checklist)


# ---------------------------------------------------------------------------
# Get checklist
# ---------------------------------------------------------------------------


@router.get(
    "/projects/{project_id}/quality/checklists/{checklist_id}",
    response_model=ChecklistSchema,
    summary="Get a single quality checklist",
)
def get_checklist(
    project_id: str,
    checklist_id: str,
    cfg: Settings = Depends(_get_settings),
) -> ChecklistSchema:
    """Return a single checklist by *checklist_id*, or 404 if absent."""
    checklist = _load_checklist_or_404(project_id, checklist_id, cfg)
    return _checklist_schema(checklist)


# ---------------------------------------------------------------------------
# Update checklist item
# ---------------------------------------------------------------------------


@router.put(
    "/projects/{project_id}/quality/checklists/{checklist_id}/items/{item_id}",
    response_model=ChecklistItemSchema,
    summary="Update a checklist item's status and notes",
)
def update_checklist_item(
    project_id: str,
    checklist_id: str,
    item_id: str,
    body: ChecklistItemUpdateRequest,
    cfg: Settings = Depends(_get_settings),
) -> ChecklistItemSchema:
    """Set a new status (and optionally notes) on *item_id* within *checklist_id*."""
    try:
        ChecklistItemStatus(body.status)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=(
                f"Invalid status '{body.status}'. "
                f"Must be one of: {[s.value for s in ChecklistItemStatus]}"
            ),
        )
    checklist = _load_checklist_or_404(project_id, checklist_id, cfg)
    item = checklist.update_item(item_id, status=body.status, notes=body.notes)
    if item is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"ChecklistItem '{item_id}' not found in project '{project_id}'",
        )
    path = _checklist_path(project_id, checklist_id, cfg.projects_root)
    checklist.save(path)
    return _item_schema(item)


# ---------------------------------------------------------------------------
# Delete checklist
# ---------------------------------------------------------------------------


@router.delete(
    "/projects/{project_id}/quality/checklists/{checklist_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a quality checklist",
)
def delete_checklist(
    project_id: str,
    checklist_id: str,
    cfg: Settings = Depends(_get_settings),
) -> None:
    """Remove the checklist file for *checklist_id* in *project_id*."""
    path = _checklist_path(project_id, checklist_id, cfg.projects_root)
    if not path.exists():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Checklist '{checklist_id}' not found in project '{project_id}'",
        )
    path.unlink()


# ===========================================================================
# Test-suite endpoints
# ===========================================================================

# ---------------------------------------------------------------------------
# List test suites  — static, must precede parameterised routes
# ---------------------------------------------------------------------------


@router.get(
    "/projects/{project_id}/quality/test-suites",
    response_model=TestSuiteListResponse,
    summary="List all test suites for a project",
)
def list_test_suites(
    project_id: str,
    cfg: Settings = Depends(_get_settings),
) -> TestSuiteListResponse:
    """Return all test suites for *project_id* (empty list if none exist yet)."""
    suites_dir = _suites_dir(project_id, cfg.projects_root)
    suites: list[TestSuiteSchema] = []
    if suites_dir.exists():
        for yaml_file in sorted(suites_dir.glob("*.yaml")):
            try:
                suite = TestSuite.load(yaml_file)
                suites.append(_suite_schema(suite))
            except Exception:
                continue
    return TestSuiteListResponse(project_id=project_id, suites=suites)


# ---------------------------------------------------------------------------
# Create test suite
# ---------------------------------------------------------------------------


@router.post(
    "/projects/{project_id}/quality/test-suites",
    response_model=TestSuiteSchema,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new test suite",
)
def create_test_suite(
    project_id: str,
    body: TestSuiteCreateRequest,
    cfg: Settings = Depends(_get_settings),
) -> TestSuiteSchema:
    """Create and persist a new TestSuite for *project_id*."""
    for case_req in body.cases:
        try:
            TestCaseType(case_req.case_type)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=(
                    f"Invalid case_type '{case_req.case_type}'. "
                    f"Must be one of: {[t.value for t in TestCaseType]}"
                ),
            )
    suite = TestSuite(
        project_id=project_id,
        feature=body.feature,
        generated_at=body.generated_at,
        spec_id=body.spec_id,
    )
    for case_req in body.cases:
        case = TestCase(
            suite_id=suite.suite_id,
            title=case_req.title,
            case_type=case_req.case_type,
            scenario=case_req.scenario,
            expected=case_req.expected,
            tags=case_req.tags,
        )
        suite.add_case(case)
    path = _suite_path(project_id, suite.suite_id, cfg.projects_root)
    suite.save(path)
    return _suite_schema(suite)


# ---------------------------------------------------------------------------
# Get test suite
# ---------------------------------------------------------------------------


@router.get(
    "/projects/{project_id}/quality/test-suites/{suite_id}",
    response_model=TestSuiteSchema,
    summary="Get a single test suite",
)
def get_test_suite(
    project_id: str,
    suite_id: str,
    cfg: Settings = Depends(_get_settings),
) -> TestSuiteSchema:
    """Return a single test suite by *suite_id*, or 404 if absent."""
    suite = _load_suite_or_404(project_id, suite_id, cfg)
    return _suite_schema(suite)


# ===========================================================================
# Analysis-report endpoints
# ===========================================================================

# ---------------------------------------------------------------------------
# List reports  — static collection endpoint
# ---------------------------------------------------------------------------


@router.get(
    "/projects/{project_id}/quality/reports",
    response_model=AnalysisReportListResponse,
    summary="List all analysis reports for a project",
)
def list_reports(
    project_id: str,
    cfg: Settings = Depends(_get_settings),
) -> AnalysisReportListResponse:
    """Return all analysis reports for *project_id* (empty list if none exist yet)."""
    reports_dir = _reports_dir(project_id, cfg.projects_root)
    reports: list[AnalysisReportSchema] = []
    if reports_dir.exists():
        for yaml_file in sorted(reports_dir.glob("*.yaml")):
            try:
                report = AnalysisReport.load(yaml_file)
                reports.append(_report_schema(report))
            except Exception:
                continue
    return AnalysisReportListResponse(project_id=project_id, reports=reports)


# ---------------------------------------------------------------------------
# Create analysis report
# ---------------------------------------------------------------------------


@router.post(
    "/projects/{project_id}/quality/reports",
    response_model=AnalysisReportSchema,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new analysis report",
)
def create_report(
    project_id: str,
    body: AnalysisReportCreateRequest,
    cfg: Settings = Depends(_get_settings),
) -> AnalysisReportSchema:
    """Create and persist a new AnalysisReport for *project_id*."""
    for finding_req in body.findings:
        try:
            SeverityLevel(finding_req.severity)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=(
                    f"Invalid severity '{finding_req.severity}'. "
                    f"Must be one of: {[s.value for s in SeverityLevel]}"
                ),
            )
    report = AnalysisReport(
        project_id=project_id,
        analyzed_at=body.analyzed_at,
        spec_id=body.spec_id,
        plan_id=body.plan_id,
    )
    for finding_req in body.findings:
        finding = AnalysisFinding(
            report_id=report.report_id,
            severity=finding_req.severity,
            category=finding_req.category,
            description=finding_req.description,
            recommendation=finding_req.recommendation,
        )
        report.add_finding(finding)
    path = _report_path(project_id, report.report_id, cfg.projects_root)
    report.save(path)
    return _report_schema(report)

File: frontend/src/api/quality.ts
/**
 * API client for the SpecForge Quality Guardian module (Module 7).
 *
 * Endpoints:
 *   GET    /api/v1/projects/{projectId}/quality/checklists
 *   POST   /api/v1/projects/{projectId}/quality/checklists
 *   GET    /api/v1/projects/{projectId}/quality/checklists/{checklistId}
 *   PUT    /api/v1/projects/{projectId}/quality/checklists/{checklistId}/items/{itemId}
 *   DELETE /api/v1/projects/{projectId}/quality/checklists/{checklistId}
 *   GET    /api/v1/projects/{projectId}/quality/test-suites
 *   POST   /api/v1/projects/{projectId}/quality/test-suites
 *   GET    /api/v1/projects/{projectId}/quality/test-suites/{suiteId}
 *   GET    /api/v1/projects/{projectId}/quality/reports
 *   POST   /api/v1/projects/{projectId}/quality/reports
 */

const BASE = '/api/v1';

// ---------------------------------------------------------------------------
// Domain types
// ---------------------------------------------------------------------------

export type ChecklistItemStatus = 'pass' | 'fail' | 'skip' | 'pending';
export type SeverityLevel = 'critical' | 'high' | 'medium' | 'low' | 'info';
export type TestCaseType = 'unit' | 'integration' | 'contract' | 'e2e';

export const CHECKLIST_STATUS_COLOR: Record<ChecklistItemStatus, string> = {
  pass: '#059669',
  fail: '#dc2626',
  skip: '#ca8a04',
  pending: '#6b7280',
};

export const CHECKLIST_STATUS_ICON: Record<ChecklistItemStatus, string> = {
  pass: '✅',
  fail: '❌',
  skip: '⏭️',
  pending: '⏳',
};

export const SEVERITY_COLOR: Record<SeverityLevel, string> = {
  critical: '#7f1d1d',
  high: '#dc2626',
  medium: '#ca8a04',
  low: '#2563eb',
  info: '#6b7280',
};

export const SEVERITY_BG: Record<SeverityLevel, string> = {
  critical: '#fef2f2',
  high: '#fef2f2',
  medium: '#fefce8',
  low: '#eff6ff',
  info: '#f3f4f6',
};

export const TEST_TYPE_COLOR: Record<TestCaseType, string> = {
  unit: '#7c3aed',
  integration: '#2563eb',
  contract: '#059669',
  e2e: '#ca8a04',
};

export const SEVERITY_ORDER: Record<SeverityLevel, number> = {
  critical: 0,
  high: 1,
  medium: 2,
  low: 3,
  info: 4,
};

export interface ChecklistItem {
  item_id: string;
  category: string;
  description: string;
  status: ChecklistItemStatus;
  notes: string;
}

export interface Checklist {
  checklist_id: string;
  project_id: string;
  title: string;
  spec_id: string | null;
  created_at: string;
  items: ChecklistItem[];
}

export interface ChecklistListResponse {
  project_id: string;
  checklists: Checklist[];
}

export interface TestCase {
  case_id: string;
  title: string;
  type: TestCaseType;
  description: string;
  acceptance_criteria: string[];
  status: 'draft' | 'active' | 'deprecated';
}

export interface TestSuite {
  suite_id: string;
  project_id: string;
  title: string;
  spec_id: string | null;
  created_at: string;
  test_cases: TestCase[];
}

export interface TestSuiteListResponse {
  project_id: string;
  suites: TestSuite[];
}

export interface AnalysisFinding {
  finding_id: string;
  severity: SeverityLevel;
  artifact_type: string;
  artifact_id: string;
  title: string;
  description: string;
  recommendation: string;
}

export interface AnalysisReport {
  report_id: string;
  project_id: string;
  created_at: string;
  summary: string;
  overall_score: number;
  findings: AnalysisFinding[];
}

export interface AnalysisReportListResponse {
  project_id: string;
  reports: AnalysisReport[];
}

// ---------------------------------------------------------------------------
// HTTP helpers
// ---------------------------------------------------------------------------

async function request<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json', ...(init?.headers ?? {}) },
    ...init,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`HTTP ${res.status}: ${text}`);
  }
  if (res.status === 204) return undefined as unknown as T;
  return res.json() as Promise<T>;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export function listChecklists(projectId: string): Promise<ChecklistListResponse> {
  return request(`${BASE}/projects/${projectId}/quality/checklists`);
}

export function createChecklist(
  projectId: string,
  body: Pick<Checklist, 'title'> & Partial<Pick<Checklist, 'spec_id'>>
): Promise<Checklist> {
  return request(`${BASE}/projects/${projectId}/quality/checklists`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function getChecklist(projectId: string, checklistId: string): Promise<Checklist> {
  return request(`${BASE}/projects/${projectId}/quality/checklists/${checklistId}`);
}

export function updateChecklistItem(
  projectId: string,
  checklistId: string,
  itemId: string,
  body: Partial<Pick<ChecklistItem, 'status' | 'notes'>>
): Promise<ChecklistItem> {
  return request(
    `${BASE}/projects/${projectId}/quality/checklists/${checklistId}/items/${itemId}`,
    { method: 'PUT', body: JSON.stringify(body) }
  );
}

export function deleteChecklist(projectId: string, checklistId: string): Promise<void> {
  return request(`${BASE}/projects/${projectId}/quality/checklists/${checklistId}`, {
    method: 'DELETE',
  });
}

export function listTestSuites(projectId: string): Promise<TestSuiteListResponse> {
  return request(`${BASE}/projects/${projectId}/quality/test-suites`);
}

export function createTestSuite(
  projectId: string,
  body: Pick<TestSuite, 'title'> & Partial<Pick<TestSuite, 'spec_id'>>
): Promise<TestSuite> {
  return request(`${BASE}/projects/${projectId}/quality/test-suites`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function getTestSuite(projectId: string, suiteId: string): Promise<TestSuite> {
  return request(`${BASE}/projects/${projectId}/quality/test-suites/${suiteId}`);
}

export function listAnalysisReports(projectId: string): Promise<AnalysisReportListResponse> {
  return request(`${BASE}/projects/${projectId}/quality/reports`);
}

export function createAnalysisReport(
  projectId: string,
  body: { summary: string }
): Promise<AnalysisReport> {
  return request(`${BASE}/projects/${projectId}/quality/reports`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

// ---------------------------------------------------------------------------
// Demo data
// ---------------------------------------------------------------------------

export function getDemoChecklist(): Checklist {
  return {
    checklist_id: 'demo-checklist-001',
    project_id: 'demo-project',
    title: 'Module 7 Quality Checklist',
    spec_id: 'spec-demo',
    created_at: '2026-04-10T08:00:00Z',
    items: [
      // Specification
      { item_id: 'i-001', category: 'Specification', description: 'All user stories have acceptance criteria', status: 'pass', notes: '' },
      { item_id: 'i-002', category: 'Specification', description: 'Non-functional requirements documented', status: 'pass', notes: '' },
      { item_id: 'i-003', category: 'Specification', description: 'Edge cases identified and documented', status: 'fail', notes: 'Missing error state scenarios for OAuth flow' },
      // Architecture
      { item_id: 'i-004', category: 'Architecture', description: 'API contracts defined for all public endpoints', status: 'pass', notes: '' },
      { item_id: 'i-005', category: 'Architecture', description: 'Data model reviewed for normalization', status: 'pass', notes: '' },
      { item_id: 'i-006', category: 'Architecture', description: 'Security threat model completed', status: 'pending', notes: '' },
      // Testing
      { item_id: 'i-007', category: 'Testing', description: 'Unit test coverage ≥ 80%', status: 'pass', notes: 'Coverage at 87%' },
      { item_id: 'i-008', category: 'Testing', description: 'Integration tests for all API endpoints', status: 'fail', notes: 'Missing tests for DELETE /sessions endpoint' },
      { item_id: 'i-009', category: 'Testing', description: 'E2E happy-path scenarios automated', status: 'skip', notes: 'Deferred to next sprint' },
      // Deployment
      { item_id: 'i-010', category: 'Deployment', description: 'CI/CD pipeline configured', status: 'pass', notes: '' },
      { item_id: 'i-011', category: 'Deployment', description: 'Environment variables documented in README', status: 'pending', notes: '' },
    ],
  };
}

export function getDemoTestSuite(): TestSuite {
  return {
    suite_id: 'demo-suite-001',
    project_id: 'demo-project',
    title: 'Core Domain Test Suite',
    spec_id: 'spec-demo',
    created_at: '2026-04-10T08:30:00Z',
    test_cases: [
      {
        case_id: 'tc-001',
        title: 'Session creation with valid payload',
        type: 'unit',
        description: 'Verify that creating a session with all required fields returns a valid session object.',
        acceptance_criteria: ['Returns session_id', 'Status is idle', 'started_at is set'],
        status: 'active',
      },
      {
        case_id: 'tc-002',
        title: 'Session status transition: idle → running',
        type: 'unit',
        description: 'Verify that a session can transition from idle to running status.',
        acceptance_criteria: ['Status updates to running', 'Timestamp recorded'],
        status: 'active',
      },
      {
        case_id: 'tc-003',
        title: 'GET /sessions returns all project sessions',
        type: 'integration',
        description: 'Verify that listing sessions for a project returns all created sessions.',
        acceptance_criteria: ['Returns array of sessions', 'Correct project_id filter applied'],
        status: 'active',
      },
      {
        case_id: 'tc-004',
        title: 'API contract: POST /sessions schema validation',
        type: 'contract',
        description: 'Validate request and response schemas match the OpenAPI specification.',
        acceptance_criteria: ['Request schema validated', 'Response schema validated', 'Error schema validated'],
        status: 'active',
      },
      {
        case_id: 'tc-005',
        title: 'User can create and execute a session end-to-end',
        type: 'e2e',
        description: 'Full user journey: create session, add task results, mark complete, view summary.',
        acceptance_criteria: ['Session created via UI', 'Tasks added and tracked', 'Summary visible', 'Status reflects completion'],
        status: 'active',
      },
      {
        case_id: 'tc-006',
        title: 'Compliance report linked to task result',
        type: 'integration',
        description: 'Verify that a compliance report is correctly linked to its parent task result.',
        acceptance_criteria: ['compliance_report_id set on task result', 'Report retrievable via result reference'],
        status: 'draft',
      },
    ],
  };
}

export function getDemoAnalysisReport(): AnalysisReport {
  return {
    report_id: 'demo-report-001',
    project_id: 'demo-project',
    created_at: '2026-04-10T09:00:00Z',
    summary: 'Overall quality is good. Three findings require attention before release.',
    overall_score: 74,
    findings: [
      {
        finding_id: 'f-001',
        severity: 'critical',
        artifact_type: 'Specification',
        artifact_id: 'spec-demo',
        title: 'Missing error handling spec for OAuth callback',
        description: 'The specification does not define behaviour when the OAuth provider returns an error code during the callback phase.',
        recommendation: 'Add a dedicated error scenario section covering provider errors, token expiry, and network failures.',
      },
      {
        finding_id: 'f-002',
        severity: 'high',
        artifact_type: 'Test Suite',
        artifact_id: 'suite-001',
        title: 'DELETE endpoint has no integration tests',
        description: 'The DELETE /sessions/{sessionId} endpoint is not covered by any integration test, leaving a gap in regression protection.',
        recommendation: 'Add at least two integration tests: one for successful deletion and one for deleting a non-existent resource (404).',
      },
      {
        finding_id: 'f-003',
        severity: 'medium',
        artifact_type: 'Architecture Plan',
        artifact_id: 'plan-demo',
        title: 'Security threat model not completed',
        description: 'The threat model section of the architecture plan is marked as pending. Auth flows and data storage have not been assessed.',
        recommendation: 'Complete the STRIDE threat model before the security review milestone.',
      },
      {
        finding_id: 'f-004',
        severity: 'low',
        artifact_type: 'Specification',
        artifact_id: 'spec-demo',
        title: 'Non-functional requirements lack measurable thresholds',
        description: 'Performance requirements state "fast response times" without defining specific latency targets.',
        recommendation: 'Update NFRs with measurable SLOs, e.g. p99 latency < 200 ms.',
      },
      {
        finding_id: 'f-005',
        severity: 'info',
        artifact_type: 'Task List',
        artifact_id: 'tasks-demo',
        title: 'E2E tests deferred to next sprint',
        description: 'The E2E test suite item was marked as skipped with a note to defer. This is documented and accepted.',
        recommendation: 'Ensure the deferred item is tracked in the next sprint backlog.',
      },
    ],
  };
}
